hoodie-common/src/main/java/com/uber/hoodie/common/table/log/avro/CompositeAvroLogReader.java

```diff
@@ -0,0 +1,147 @@
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *           http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package com.uber.hoodie.common.table.log.avro;
+
+import com.uber.hoodie.common.table.log.HoodieLogFile;
+import com.uber.hoodie.common.util.FSUtils;
+import com.uber.hoodie.exception.HoodieIOException;
+import org.apache.avro.Schema;
+import org.apache.avro.generic.GenericRecord;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+
+import java.io.IOException;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+/**
+ * CompositeAvroLogReader reads all versions of the logs for a given fileId.
+ * It gives a iterator which iterates through all the versions and the list of blocks for that specific version
+ * Useful for merging records in RecordReader and compacting all the delta versions
+ *
+ * @see AvroLogReader
+ */
+public class CompositeAvroLogReader {
+    private final Map<Integer, AvroLogReader> readers;
+
+    public CompositeAvroLogReader(Path partitionPath, String fileId, FileSystem fs,
+        Schema readerSchema, String logFileExtension) throws IOException {
+        Stream<HoodieLogFile> allLogFiles =
+            FSUtils.getAllLogFiles(fs, partitionPath, fileId, logFileExtension);
+        this.readers = allLogFiles.map(hoodieLogFile -> {
+            try {
+                return new AvroLogReader(hoodieLogFile, fs, readerSchema);
+            } catch (IOException e) {
+                throw new HoodieIOException(
+                    "Could not read avro records from path " + hoodieLogFile);
+            }
+        }).collect(Collectors.toMap(new Function<AvroLogReader, Integer>() {
+            @Override
+            public Integer apply(AvroLogReader avroLogReader) {
+                return avroLogReader.getFile().getLogVersion();
+            }
+        }, Function.identity()));
+    }
+
+    public Iterator<GenericRecord> readBlocks(Map<Integer, List<Long>> filesToOffsetMap)
+        throws IOException {
+        return new Iterators(filesToOffsetMap, readers);
+    }
+
+    public void close() throws IOException {
+        readers.values().forEach(s -> {
+            try {
+                s.close();
+            } catch (IOException e) {
+                throw new HoodieIOException("Unable to close " + s.getFile(), e);
+            }
+        });
+    }
+
+    public class Iterators implements Iterator<GenericRecord> {
```

Comment:
Just a test that tests some happy paths and degenerate paths 

- Just one iterator
- 0 iterators 
- iterators with 0 elements, 1 elements chained here

As long as we have some tests around this class standalone, it will be nice